pkg/prober/prober.go

```diff
@@ -0,0 +1,153 @@
+	p.componentMtx.Lock()
+	defer p.componentMtx.Unlock()
+	return p.component
+}
+
+// NewProber returns Prober reprezenting readiness and healthiness of given component.
+func NewProber(component component.Component, logger log.Logger) *Prober {
+	initialErr := fmt.Errorf(initialErrorText, component)
+	prober := &Prober{}
+	prober.SetComponent(component)
+	prober.SetLogger(logger)
+	prober.SetNotHealthy(initialErr)
+	prober.SetNotReady(initialErr)
+	return prober
+}
+
+// HandleInMux registers readiness and liveness probes to mux.
+func (p *Prober) RegisterInRouter(router *route.Router) {
+	router.Get(healthyEndpointPath, p.probeHandlerFunc(p.IsHealthy, "healthy"))
+	router.Get(readyEndpointPath, p.probeHandlerFunc(p.IsReady, "ready"))
+}
+
+// RegisterInMux registers readiness and liveness probes to mux.
+func (p *Prober) RegisterInMux(mux *http.ServeMux) {
+	mux.HandleFunc(healthyEndpointPath, p.probeHandlerFunc(p.IsHealthy, "healthy"))
+	mux.HandleFunc(readyEndpointPath, p.probeHandlerFunc(p.IsReady, "ready"))
+}
+
+func (p *Prober) writeResponse(w http.ResponseWriter, probeFunc func() error, probeType string) {
+	err := probeFunc()
+	if err != nil {
+		http.Error(w, fmt.Sprintf(errorProbeText, p.getComponent(), probeType, err), probeErrorHTTPStatus)
+		return
+	}
+	if _, e := io.WriteString(w, fmt.Sprintf(okProbeText, p.getComponent(), probeType)); e == nil {
+		level.Error(p.getLogger()).Log("msg", "failed to write probe response", "probe type", probeType, "err", err)
+	}
+}
+
+func (p *Prober) probeHandlerFunc(probeFunc func() error, probeType string) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, _ *http.Request) {
+		p.writeResponse(w, probeFunc, probeType)
+	}
+}
+
+// IsReady returns error if component is not ready and nil if it is.
+func (p *Prober) IsReady() error {
+	p.readyMtx.Lock()
+	defer p.readyMtx.Unlock()
+	return p.readiness
+}
+
+// SetReady sets components status to ready.
+func (p *Prober) SetReady() {
+	level.Debug(p.getLogger()).Log("msg", "changing probe status", "status", "ready")
+	p.SetNotReady(nil)
+}
+
+// SetNotReady sets components status to not ready with given error as a cause.
+func (p *Prober) SetNotReady(err error) {
+	p.readyMtx.Lock()
+	defer p.readyMtx.Unlock()
+	if err != nil {
+		level.Debug(p.getLogger()).Log("msg", "changing probe status", "status", "not-ready", "reason", err)
+	}
+	p.readiness = err
+}
+
+// IsHealthy returns error if component is not healthy and nil if it is.
+func (p *Prober) IsHealthy() error {
+	p.healthyMtx.Lock()
+	defer p.healthyMtx.Unlock()
+	return p.healthiness
+}
+
+// SetHealthy sets components status to healthy.
+func (p *Prober) SetHealthy() {
+	level.Debug(p.getLogger()).Log("msg", "changing probe status", "status", "healthy")
+	p.SetNotHealthy(nil)
+}
+
+// SetNotHealthy sets components status to not healthy with given error as a cause.
+func (p *Prober) SetNotHealthy(err error) {
+	p.healthyMtx.Lock()
+	defer p.healthyMtx.Unlock()
```

Comment:
Maybe add log.Warn(..).Log("msg","changing probe status", "status", "unhealthy") when the status changed from healthy to unhealthy?

Reply:
You mean dropping the debug log or debug only if status does not change and when it changes log it as warning?

Reply:
I mean maybe it's useful to log warnings when the status changes from healthy to unhealthy?

Reply:
Just a thought, feel free to ignore this :P

Reply:
Ok, I changed it as you suggested so now:
 - if status actually changes to healthy/ready it's logged as `Info` (if already healthy/ready nothing is logged)
 - if status changes to notReady/notHealthy it's logged as `Warn` with the error which caused it (if already notReady/notHealthy nothing is logged)

Is that ok ?

Reply:
awesome! :tada: